language/move-lang/src/expansion/translate.rs

```diff
@@ -317,84 +369,216 @@ fn script_(context: &mut Context, pscript: P::Script) -> E::Script {
-            )])
+fn use_(context: &mut Context, acc: &mut AliasMap, u: P::Use) {
+    let unbound_module = |mident: &ModuleIdent| -> Error {
+        vec![(
+            mident.loc(),
+            format!("Invalid 'use'. Unbound module: '{}'", mident),
+        )]
+    };
+    macro_rules! add_module_alias {
+        ($ident:expr, $alias_opt:expr) => {{
+            let alias: Name = $alias_opt.unwrap_or_else(|| $ident.0.value.name.0.clone());
+            if &alias.value == ModuleName::SELF_NAME {
+                context.restricted_self_error("module alias", &alias);
+                return;
+            }
+
+            if let Err(old_loc) = acc.add_module_alias(alias.clone(), $ident) {
+                duplicate_module_alias(context, old_loc, alias)
+            }
+        }};
+    };
+    match u {
+        P::Use::Module(mident, alias_opt) => {
+            if !context.module_members.contains_key(&mident) {
+                context.error(unbound_module(&mident));
+                return;
+            };
+            add_module_alias!(mident, alias_opt.map(|m| m.0))
+        }
+        P::Use::Members(mident, sub_uses) => {
+            if !context.module_members.contains_key(&mident) {
+                context.error(unbound_module(&mident));
+                return;
+            };
+
+            for (member, alias_opt) in sub_uses {
+                if member.value == ModuleName::SELF_NAME {
+                    add_module_alias!(mident.clone(), alias_opt);
+                    continue;
+                }
+
+                // check is member
+                let members = &context.module_members[&mident];
+                let mloc = members.loc;
+                let is_member = members.is_member(&member);
+                let is_struct = members.is_struct(&member);
+                if !is_member {
+                    let msg = format!(
+                        "Invalid 'use'. Unbound member '{}' in module '{}'",
+                        member, mident
+                    );
+                    context.error(vec![
+                        (member.loc, msg),
+                        (mloc, format!("Module '{}' declared here", mident)),
+                    ]);
+                    continue;
+                }
+
+                // if struct member, check valid struct name
+                let struct_name_err = match alias_opt.as_ref() {
+                    // No explicit alias so nothing to check
+                    None => Ok(()),
+                    // Not a struct, nothing to check
+                    Some(_) if !is_struct => Ok(()),
+                    Some(alias) => {
+                        check_valid_struct_or_constant_name(alias, "struct alias", "Struct alias")
+                    }
+                };
+                if let Err(e) = struct_name_err {
+                    // TODO point to struct declaration
+                    context.error(e);
+                    continue;
+                }
```

Comment:
Lifting this `if let` into the last branch of the match might make things a bit cleaner.